File: 2023/Day1.cs
class Day1
{
    static void Main()
    {
        List<string> data = new List<string>();
        StreamReader reader = new StreamReader("input.txt");
        string? line = reader.ReadLine();
        while (line != null)
        {
            data.Add(line);
            line = reader.ReadLine();
        }
        reader.Close();

        Console.WriteLine("Part 1 Answer: " + Part1(data));
        Console.WriteLine("Part 2 Answer: " + Part2(data));
    }

    static int Part1(List<string> data)
    {
        int total = 0;
        foreach (string line in data)
        {
            int start = 0;
            int end = line.Length - 1;
            while (!char.IsNumber(line[start]))
            {
                start++;
            }
            while (!char.IsNumber(line[end]))
            {
                end--;
            }
            total += int.Parse(line[start] + "" + line[end]);
        }
        return total;
    }

    static int Part2(List<string> data)
    {
        string[] numWords = new string[] { "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
        int total = 0;
        foreach (string line in data)
        {
            int start = 0;
            int end = line.Length - 1;
            int startDigit = -1;
            int endDigit = -1;
            while (startDigit < 0)
            {
                if (char.IsNumber(line[start]))
                {
                    startDigit = int.Parse(line[start].ToString());
                }
                else
                {
                    bool foundWord = false;
                    for (int i = 0; i < numWords.Length; i++)
                    {
                        if (start + numWords[i].Length <= line.Length
                                && line.Substring(start, numWords[i].Length) == numWords[i])
                        {
                            startDigit = i + 1;
                            foundWord = true;
                            break;
                        }
                    }
                    if (!foundWord)
                    {
                        start++;
                    }
                }
            }
            while (endDigit < 0)
            {
                if (char.IsNumber(line[end]))
                {
                    endDigit = int.Parse(line[end].ToString());
                }
                else
                {
                    bool foundWord = false;
                    for (int i = 0; i < numWords.Length; i++)
                    {
                        if (end - numWords[i].Length >= 0
                                && line.Substring(end - numWords[i].Length + 1, numWords[i].Length) == numWords[i])
                        {
                            endDigit = i + 1;
                            foundWord = true;
                            break;
                        }
                    }
                    if (!foundWord)
                    {
                        end--;
                    }
                }
            }
            total += startDigit * 10 + endDigit;
        }
        return total;
    }
}

File: 2024/Day1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <list>
#include <map>

int part1(const std::vector<std::string> &input)
{
    std::vector<int> left;
    std::vector<int> right;
    for (std::string line : input)
    {
        std::string str = "";
        for (char c : line)
        {
            if (c == ' ') {
                if (str != "") {
                    left.push_back(std::stoi(str));
                }
                str = "";
            }
            else {
                str += c;
            }
        }
        right.push_back(std::stoi(str));
    }
    sort(left.begin(), left.end());
    sort(right.begin(), right.end());
    int total = 0;
    for (int i; i < left.size(); i++)
    {
        total += abs(left[i] - right[i]);
    }
    return total;
}

int part2(const std::vector<std::string> &input)
{
    std::vector<int> left;
    std::vector<int> right;
    for (std::string line : input)
    {
        std::string str = "";
        for (char c : line)
        {
            if (c == ' ') {
                if (str != "") {
                    left.push_back(std::stoi(str));
                }
                str = "";
            }
            else {
                str += c;
            }
        }
        right.push_back(std::stoi(str));
    }
    std::map<int, int> freq;
    for (int r : right)
    {
        freq[r]++;
    }
    int total = 0;
    for (int l : left)
    {
        total += l * freq[l];
    }
    return total;
}

int main()
{
    std::ifstream file("input.txt");
    std::vector<std::string> input;
    std::string line;
    while (std::getline(file, line))
    {
        input.push_back(line);
    }
    file.close();

    fprintf(stdout, "Part 1 Answer: %d\n", part1(input));
    fprintf(stdout, "Part 2 Answer: %d\n", part2(input));

    return 0;
}